crypto/crypto/src/lib.rs

```diff
@@ -3,6 +3,7 @@
 
 #![forbid(unsafe_code)]
 #![deny(missing_docs)]
+#![cfg_attr(mirai, allow(incomplete_features), feature(const_generics))]
```

Comment:
This is bound to scare people off: `const_generics` is something we do not want anywhere near production releases, and we really want to make it really easy to check that fact.

Here's one way I think this could be achieved: 
- having a *crate* feature in the crypto crate, 
- making it non-default, and 
- making it the condition for compilation of the distinct module that 
    + introduces the tag types, 
    + and therein turns on `allow(incomplete_features)` and `#![feature(const_generics)]`.

Then ideally the `has_tag` preconditions refer to the fully-qualified name for this module.

Reply:
It seems that `#![feature(const_generics)]` cannot be used at module level. I've tried to move the definitions of tag types to a separate tag crate, but I'm not sure how to make the tag crate only accessible to the libra-crypto crate.

Reply:
Have you tried conditional compilation of the feature gate?

`#![cfg_attr(all(feature = "mirai-analysis-on", [debug_assertions](https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions)), feature(link_llvm_intrinsics))]`

Reply:
what's wrong with const_generics?

Reply:
Nothing much, other than that is not stable. We are not using any strange features and the standard libraries are using constant generics as well. I expect that something close enough like this will become part of the stable language at some point.

Reply:
`const_generics` is not only unstable, it's [incomplete](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_feature/constant.INCOMPLETE_FEATURES.html). It's nowhere near stabilization. 

I'd urge you to consider the risk of allowing anybody to depend on the activation of such a feature (which, as it stands in the current state of the PR, is crate-wide).  In fact I'm going to formally request changes to reflect that.

Reply:
Given we build and run all of libra using a stable compiler, we can have some solace in the fact that if that feature happens to get turned on somehow, that compilation should fail.

Reply:
Just to help me understand the issue better: With the code as it is in this PR, can the stable compiler compile Libra?

If so, what is wrong with the code as it is in this PR?

Reply:
Given the current [CI failures](https://circleci.com/gh/libra/libra/174477?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) it seems like there's probably some more work to be done, specifically to ensure that our CI does not try to build/test with anything that enables this feature.

Reply:
Indeed, the only way code that depends on this feature should ever be compiled is when the mirai flag is set, which is something the CI should never do.

Reply:
The current CI failure comes from the fact that tag type definitions are moved to a separate crate in the latest commit.

Originally, the tag types were defined inside the libra-crypto crate, so there was no CI failure.

Reply:
It seems to me that doing things in the original way makes sense. I would be in favor of reverting to it.

Reply:
Reverted back to the original way, but moved tag definitions to a separate tags module, which gets compiled only if both mirai and debug_assestions are on.

Is is still necessary to have a crate feature like `mirai-tag-analysis`, rather than use the mirai identifier directly?

Reply:
- the use of the `mirai` feature is fine,
- after having played around with the code, the current submodule for tags is great,
- I'd still love a comment above the `cfg_attr(mirai, allow(incomplete_features), ...` which states something like:
```
//! This feature gets turned on only if libra-crypto is compiled via MIRAI in a nightly build.
```